### ISSUE TITLE
[1.30.4] Back is pressed instead of keyboard hidden if it's performed too quickly

### ISSUE THREAD

**Issue by LukaszCuler** [mobile-dev-inc/maestro#1303]
**Describe the bug**
Let's open a screen with EditText, which is focused automatically. Then, instantly hide the on-screen keyboard with `hideKeyboard` command. Usually, it works fine - we end up with focused EditText which we can feed with text and with complete visibility of the rest of the screen. However, sometimes `hideKeyboard` performs `back` before the keyboard is fully visible which navigates the app to the previous screen, so it's kind of a race condition.

**Expected behavior**
`hideKeyboard` should ensure that the on-screen keyboard is visible before trying to hide it to reduce flakiness.

**Environment information (please complete the following information):**

- Maestro version: [1.30.4]
- Platform: [Android 13]
- Framework: [Native Android]
- Device model and OS version: Android Emulator based on Android 13
- Host: Mac M1

**Comment by arapocket** [comment]
This is still an issue right? Think I'm running into it.

**Comment by Fishbowler** [comment]
@arapocket Could be, it's not one I've looked into.

In the meantime, some workaround ideas:
- add a `waitForAnimationToEnd` before `hideKeyboard`
- tap something non-interactive instead of using `hideKeyboard`, like a header. That would dismiss the keyboard without risking unexpected navigation.

**Comment by alexbaileyuk** [comment]
This still appears to be an issue with the following configuration:

Maestro version: [2.0.8]
Platform: [Android 16]
Framework: [Expo / React Native]
Device model and OS version: Android Emulator based on Android 16
Host: Windows

`waitForAnimationToEnd` did not work for me. However, using Maestro Studio, I found a couple of hacks:

```yaml
- tapOn:
    id: "android:id/input_method_nav_back"

# OR

- tapOn: "Back"

# OR

- tapOn:
    point: "10%,98%"
```

<img width="875" height="452" alt="Image" src="https://github.com/user-attachments/assets/f7f2a126-5959-4e72-9a5e-8ad31a7fb496" />

Of course, your mileage may vary. These are all hacks.

**Comment by alexbaileyuk** [comment]
Unfortunately I'm having further trouble on Android devices where stylus input is enabled. Specifically "Write in text fields".

<img width="1280" height="2856" alt="Image" src="https://github.com/user-attachments/assets/6e22e331-9180-4564-a6e1-f1e9b06f4757" />

If I enable "Show on-screen keyboard", then the keyboard always shows up. Great. If I disable "Use stylus to write in text fields" then it doesn't show the stylus input by default.

I've tried to extend the above further to handle test devices with a stylus input which don't show the keyboard unless you change some system settings. I ended up with the below as `safe-dismiss-keyboard.yaml`:

```yaml
appId: ai.xxxxxxxxx.mobile.dev
---
# When the stylus input method is active, a sidebar is shown (not a keyboard)
- runFlow:
    when:
      visible:
        id: 'com.google.android.inputmethod.latin:id/0_resource_name_obfuscated'
    commands:
      - back

# When the stylus input method is active, or a password input is active, a keyboard is shown
# https://github.com/mobile-dev-inc/maestro/issues/1303
- runFlow:
    when:
      visible:
        id: 'android:id/input_method_nav_back'
    commands:
      - hideKeyboard

- assertNotVisible:
    id: 'com.google.android.inputmethod.latin:id/0_resource_name_obfuscated'

- assertNotVisible:
    id: 'android:id/input_method_nav_back'
```

I'm not totally sure how reliable this is. It definitely slows down execution considerably.

**Comment by Fishbowler** [comment]
@alexbaileyuk I suspect what you're describing is distinct from the issue, although I've read it a couple of times and I'm not sure. Did you want stylus input or not? If not, why is it on?

**Comment by alexbaileyuk** [comment]
@Fishbowler you might be right, I'm really just updating the previous shared work around that I had for the "back" being pressed if it's done early. By using `runFlow` it seems to wait for a moment to see if the element is visible before dismissing the keyboard. 

I'll open another issue if I can't get round the stylus input issue. It's enabled by default on a clean emulator. The workaround above should still be viable for the originally mentioned issue (at least, I cannot reproduce the issue using the above). A caveat to it is as a hacky workaround is the stylus input.

**Comment by mhammadjaber00** [comment]
Opened https://github.com/mobile-dev-inc/maestro/pull/3605 for this.

The race described here is one way in, but the trigger is broader: `AndroidDriver.hideKeyboard()` sends `input keyevent 4` unconditionally, so *any* `hideKeyboard` that runs while no IME is on screen navigates back instead. That makes it bite on conditional paths and shared subflows that cannot know whether a field is focused, not just on auto-focus timing.

`Orchestra.hideKeyboardCommand()` already calls `isKeyboardVisible()` — but only after the dismiss, to raise `HideKeyboardFailure`. The PR asks the same question first and returns early when there is nothing to dismiss, which is what this issue asks for. It also covers the no-keyboard case of #1373, since the guard sits above the driver.
